### PR TITLE
Add ink! e2e-tests

### PR DESCRIPTION
**Pull request by prxgr4mm3r**

### CODE REVIEW

File: src/commands/init/index.ts [hunk 1]
```diff
@@ -269,6 +269,15 @@ export class Init extends SwankyCommand<typeof Init> {
       runningMessage: "Copying contract template files",
     });
 
+    this.taskQueue.push({
```
Comment (ipapandinas) [inline src/commands/init/index.ts:272]: This task needs to be perform only for `contractTemplate=psp22`

File: src/commands/contract/test/index.ts [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+      }
+
+      console.log(`Testing contract: ${contractName}`);
+
+      if (!flags.mocha) {
+        await spinner.runCommand(
+          async () => {
+            return new Promise<string>((resolve, reject) => {
+              const compileArgs = [
+                "test",
```
Comment (ipapandinas) [inline src/commands/contract/test/index.ts:79]: We should be able to run e2e tests against local swanky node by providing the following env variable: `CONTRACTS_NODE="LOCAL_SWANKY_NODE_PATH"`. This is optional and can be handle by an optional boolean `local-node`.

File: src/commands/contract/test/index.ts [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+export class TestContract extends SwankyCommand<typeof TestContract> {
+  static description = "Run tests for a given contact";
+
+  static flags = {
+    all: Flags.boolean({
+      default: false,
+      char: "a",
+      description: "Set all to true to compile all contracts",
+    }),
+    mocha: Flags.boolean({
```
Comment (ipapandinas) [inline src/commands/contract/test/index.ts:27]: Not sure about this flag, this initial idea was to define it during init:

> I suggest in the init phase, if ink! is selected, only bootstrap ink! e2e, and mocha for ask!

**Review by ipapandinas — CHANGES_REQUESTED** [review]

File: src/commands/contract/test/index.ts (hunk 3 above)
Comment (prxgr4mm3r) [reply src/commands/contract/test/index.ts:27]: @ipapandinas  Am I understand right?  We want to pick what type of tests we will use when we add a contract or init project.

File: src/commands/contract/test/index.ts (hunk 2 above)
Comment (prxgr4mm3r) [reply src/commands/contract/test/index.ts:91]: I made it, but e2e-tests are not working on swanky-node for some reason

File: src/commands/contract/test/index.ts [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+    if (args.contractName === undefined && !flags.all) {
+      throw new InputError("No contracts were selected to compile");
+    }
+
+    const contractNames = flags.all
+      ? Object.keys(this.swankyConfig.contracts)
+      : [args.contractName];
+
+    const testDir = path.resolve("tests");
```
Comment (ipapandinas) [inline src/commands/contract/test/index.ts:56]: Move this to `--mocha` and test it throwing a FileError if no "tests" folder exists.

File: src/commands/contract/test/index.ts (hunk 3 above)
Comment (ipapandinas) [reply src/commands/contract/test/index.ts:31]: We can preserve the `--mocha` (check the global comment for UXcontext). However perform `generateTypes` in that case.

File: src/commands/contract/test/generate.ts
Comment (ipapandinas) [inline src/commands/contract/test/generate.ts:1]: Implement this command as a `generate type` subcommand in *src/commands/generate/types.ts*

File: src/commands/contract/test.ts
Comment (ipapandinas) [inline src/commands/contract/test.ts:1]: Preserve this subcommand

**Review by ipapandinas — CHANGES_REQUESTED** [review]
The goal is to set up the correct testing strategy based on the development dependencies used:
- ink! end-to-end (E2E) testing for ink! smart contracts.
- Mocha TypeScript (TS) testing for ask! smart contracts.

Since our current contract templates are for ink! contracts, there is no need to copy TS tests with `copyContractTemplateFiles` on the `init` (generate only) command and on the `contract new` command. Only E2E test helpers are needed for **psp22** contracts template.

The tricky part is on the `init -c` (convert) command when converting an existing project into a Swanky project. Then, we need to first detect the development dependencies used.
 
However, to be honest, I don't see comprehensive support for ask! in Swanky. The [Astar documentation](https://docs.astar.network/docs/build/wasm/ask_contracts/) recommends using ink!. Also this [issue](https://github.com/inkdevhub/swanky-cli/issues/132) mention that:

> If user choses ink! as a language, tests should be scaffolded using the ink! e2e framework. Using mocha (node.js) is redundant.

@prxgr4mm3r I suggest removing scaffolding for TS tests by default in the following steps:

1. Refactor `copyContractTemplateFiles` by removing the test copy.
2. Refactor `copyTestHelpers`, renaming it to `prepareTestFiles` and accepting a "type" argument of type "e2e" or "mocha". On `init` (generate only) and on the `contract new` command, it runs as `prepareTestFiles("e2e")` for **psp22** templates.
3. Preserve the previous folder structure for tests with `contract test` (no need for a test folder with an index.ts). Preserve the actual optional `--mocha` flag to run TS tests but perform a `generateTypes`. Otherwise, E2E tests run by default.
4. Create a dedicated generate subcommand folder containing 2 commands: 

- `generate types` (previous `TypegenCommand`)
- `generate test` with contract name argument. By default, it performs `prepareTestFiles("e2e")` unless an optional template is passed as flag, then it scaffolds TS tests in the selected contract test folder with prepareTestFiles("mocha"). This is similar to the actual test copy in `copyContractTemplateFiles`. In both cases, if a "tests" folder already exists when running `generate test`, prompt a warning confirmation to the user.

File: src/commands/init/index.ts [hunk 5]
```diff
@@ -544,7 +553,7 @@ async function detectTests(pathToExistingProject: string): Promise<string | unde
 }
 
 async function readRootCargoToml(pathToProject: string) {
-  const rootTomlPath = path.resolve(pathToProject, "Cargo.toml");
+  const rootTomlPath = path.resolve(pathToProject, "Cargo.toml.hbs.hbs");
```
Comment (ipapandinas) [inline src/commands/init/index.ts:556]: Why did you add the ".hbs.hbs" extension in the convert section of the `init` command? @prxgr4mm3r

Comment (prxgr4mm3r) [reply src/commands/init/index.ts]: I think my IDE did it automatically for some reason. Already fixed it.

**Review by ipapandinas — APPROVED** [review]
LGTM

**Review by pmikolajczyk41 — APPROVED** [review]

File: src/commands/contract/compile.ts [hunk 6]
```diff
@@ -108,12 +108,6 @@ export class CompileContract extends SwankyCommand<typeof CompileContract> {
       await spinner.runCommand(async () => {
         return storeArtifacts(artifactsPath, contractInfo.name, contractInfo.moduleName);
       }, "Moving artifacts");
-
-      await spinner.runCommand(
-        async () => await generateTypes(contractInfo.name),
-        `Generating ${contractName} contract ts types`,
-        `${contractName} contract's TS types Generated successfully`
-      );
```
[the remark this answers is not in the document]
Comment (ipapandinas) [reply src/commands/contract/compile.ts:116]: TS types were previously used in our Mocha testing strategy. However, as we shift towards default E2E testing, we've removed them from the compile command. For users who still wish to use TS with Mocha testing, we've introduced a specific `swanky generate types` command

File: src/templates/test_helpers/lib.rs.hbs [hunk 7]
```diff
@@ -0,0 +1,18 @@
+#[macro_export]
+macro_rules! address_of {
+    ($account:ident) => {
+        ink_e2e::account_id(ink_e2e::AccountKeyring::$account)
```
[the remark this answers is not in the document]
Comment (ipapandinas) [reply src/templates/test_helpers/lib.rs.hbs:4]: yes, nice suggestion

File: src/templates/contracts/psp22/contract/Cargo.toml.hbs [hunk 8]
```diff
@@ -10,7 +10,11 @@ ink = { version = "4.2.1", default-features = false}
 scale = { package = "parity-scale-codec", version = "3", default-features = false, features = ["derive"] }
 scale-info = { version = "2.6", default-features = false, features = ["derive"], optional = true }
 
-openbrush = { git = "https://github.com/Brushfam/openbrush-contracts", tag = "4.0.0-beta", default-features = false, features = ["psp22"] }
+openbrush = { git = "https://github.com/Brushfam/openbrush-contracts", tag = "4.0.0", default-features = false, features = ["psp22"] }
```
[the remark this answers is not in the document]
Comment (ipapandinas) [reply src/templates/contracts/psp22/contract/Cargo.toml.hbs:13]: Yes, integrating this feature is part of our upcoming roadmap for the inkdevhub 2.0 proposal. I'm considering removing the openbrush PSP22 templates for this release also.

File: src/templates/test_helpers/lib.rs.hbs [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+    };
+}
+
+#[macro_export]
+macro_rules! balance_of {
+    ($contract_ref:ident, $client:ident, $address:ident, $account:ident) => \{{
```
[the remark this answers is not in the document]
Comment (ipapandinas) [reply src/templates/test_helpers/lib.rs.hbs:10]: @prxgr4mm3r

[the remark this answers is not in the document]
Comment (prxgr4mm3r) [reply src/templates/test_helpers/lib.rs.hbs:10]: @pmikolajczyk41 [Handlebars](https://handlebarsjs.com/guide/#what-is-handlebars) parser throws an error if I don't use '\\' before '{{' because it expects `{{value_that_passed_when_the_template_is_processing}}` instead.

[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply src/templates/test_helpers/lib.rs.hbs:10]: oh, okay, it's hbs file - my bad; thanks for pointing this out!

**Review by ipapandinas — APPROVED** [review]
